File: libs/backend/src/models/game/game.ts
import mongoose, { Schema, Document } from "mongoose";
import { GAME, PUZZLE, SUBMISSION, USER } from "../../utils/constants/model.js";
import { DEFAULT_GAME_LENGTH_IN_MILLISECONDS, GameEntity } from "types";
import gameOptionsSchema from "./game-config.js";

export interface GameDocument extends Document, GameEntity {}

const gameSchema = new Schema<GameDocument>({
	createdAt: {
		default: Date.now,
		type: Date
	},
	endTime: {
		default: () => Date.now() + DEFAULT_GAME_LENGTH_IN_MILLISECONDS,
		required: true,
		type: Date
	},
	options: gameOptionsSchema,
	owner: {
		ref: USER,
		required: true,
		type: mongoose.Schema.Types.ObjectId
	},
	playerSubmissions: [
		{
			ref: SUBMISSION,
			required: false,
			type: Schema.Types.ObjectId
		}
	],
	players: [
		{
			ref: USER,
			required: true,
			type: Schema.Types.ObjectId
		}
	],
	
	puzzle: {
		ref: PUZZLE,
		required: true,
		type: Schema.Types.ObjectId
	},
	startTime: {
		default: Date.now,
		required: true,
		type: Date
	},
});

const Game = mongoose.model<GameDocument>(GAME, gameSchema);
export default Game;

File: libs/elixir-backend/codin_cod/lib/codin_cod/game.ex
defmodule CodinCod.Game do
  use Ecto.Schema
  import Ecto.Changeset
  alias CodinCod.Game.GameOptions
  alias CodinCod.Accounts.User
  alias CodinCod.Puzzle


  schema "games" do
    belongs_to :puzzle, Puzzle
    belongs_to :owner, User
    many_to_many :players, User, join_through: "games_players"

    field :start_time, :utc_datetime
    embeds_one :options, GameOptions

# export const gameEntitySchema = z.object({
# 	players: z.array(objectIdSchema.or(userDtoSchema)),
# 	owner: objectIdSchema.or(userDtoSchema),
# 	puzzle: objectIdSchema.or(puzzleDtoSchema),
# 	startTime: acceptedDateSchema,
# 	endTime: acceptedDateSchema,
# 	options: gameOptionsSchema,
# 	createdAt: acceptedDateSchema,
# 	playerSubmissions: z
# 		.array(objectIdSchema.or(submissionDtoSchema))
# 		.default([]),
# });

    timestamps(type: :utc_datetime)
  end

  @doc false
  def changeset(game, attrs) do
    game
    |> cast(attrs, [])
    |> validate_required([])
  end

  defmodule GameOptions do
    use Ecto.Schema
    alias CodinCod.ProgrammingLanguage

    @required_fields ~w(duration_in_seconds)a
    @optional_fields ~w(visibility mode)a

    embedded_schema do
      many_to_many(:allowed_languages, ProgrammingLanguage, join_through: "game_options_programming_languages")

      field(:duration_in_seconds, :integer)
      field(:visbility, Ecto.Enum, values: ~w(public private)a, default: :public)
      field(:mode, Ecto.Enum, values: ~w(casual rated)a, default: :rated)
    end

    @doc false
    def changeset(profile_info, attrs) do
      profile_info
      |> cast(attrs, @required_fields ++ @optional_fields)
      |> cast_assoc(:allowed_languages)
      |> validate_required(@required_fields)
    end
  end
end
